README.md
# then-ajax

## 安装
```bash
npm install then-ajax --save
```
## 使用
```javascript
import {thenAjax, thenPost, thenGet, thenJsonp, IAjaxOptions} from 'then-ajax'
```

## api
### thenAjax(opts): Promise
- `opts.url` (string) 请求 URL
- `opts.data` (object) 数据 data, optional
- `opts.methods` (string) 请求类型 默认值 `get`, optional
- `opts.contentType` (string) header content-type, optional
- `opts.async` (boolean) async, optional
- `opts.timeout` (number) 设置超时时间, optional
- `opts.before` (function) 请求前执行函数, optional
- `opts.jsonp` (string) 设置 jsonp 默认值为 `callback`, optional
- `opts.jsonpcb` (string) 设置 jsonpcb 名称, optional

### thenGet(url, data, opts): Promise
- `url` (string) 请求 URL
- `data` (object) 数据 data, optional
- `opts` (object) 同上，略, optional

### thenPost(url, data, opts): Promise
- `url` (string) 请求 URL
- `data` (object) 数据 data, optional
- `opts` (object) 同上，略, optional

### thenJsonp(url, data, opts): Promise
- `url` (string) 请求 URL
- `data` (object) 数据 data, optional
- `opts` (object) 同上，略, optional

## example
```javascript
import syncJsonp from 'sync-jsonp';

thenAjax({
  dataType: 'json',
  url: 'http://hostname.com/path/to/api'
}).then((data) => {
  console.log('===', data);
});

thenAjax({
  data: {
    a: 2
  },
  dataType: 'json',
  url: 'http://hostname.com/path/to/api',
}).then((data) => {
  console.log('===', data);
});
```



history.md
# 历史版本
## 0.2.0(2020-01-11)
* feat: 调整格式让组件支持 tree-shaking

## 0.1.2(2019-07-28)
* feat: 添加 `timeout` 自动 抛出 new Error('timeout') 处理

## 0.1.1(2019-07-01)
* del: 去掉 `export default` 以兼容 ie8

## 0.1.0(2019-07-01)
* feat: 诞生


test/src/entry/index/index.ts
import {thenAjax, thenJsonp } from '../../../../output/then-ajax';
import wDemo from '../../components/widget/demo/demo';
import './index.scss';
wDemo();

thenAjax({
  dataType: 'json',
  timeout: 1,
  url: 'http://www.yy.com/yyweb/user/queryUserInfo.json'
}).then((data) => {
  console.log('01 thenAjax', data);
}).catch((er) => {
  console.log('01 error', er.message);
});

thenAjax({
  data: {
    a: 2
  },
  dataType: 'json',
  url: 'http://www.yy.com/yyweb/user/queryUserInfo.json'
}).then((data) => {
  console.log('02 thenAjax', data);
}).catch((er) => {
  console.log('02 error', er);
});

thenJsonp('http://www.yy.com/yyweb/user/queryUserInfo.json').then((data) => {
  console.log('03 thenJsonp', data);
}).catch((er) => {
  console.log('03 error', er);
});

thenJsonp('http://www.yy.com/yyweb/user/queryUserInfo.json', undefined, { timeout: 1}).then((data) => {
  console.log('04 thenJsonp', data);
}).catch((er) => {
  console.log('04 error', er);
});

// thenJsonp(
//   'http://www.yy.com/yyweb/user/queryUserInfo.json',
//   undefined,
//   { jsonp: 'abc', jsonpcb: 'zxc' }
// ).then((data) => {
//   console.log('===', 'jsonp', data);
// });


rollup.config.js
import pkg from './package.json';
import typescript from 'rollup-plugin-typescript2';
import commonjs from 'rollup-plugin-commonjs';
import nodeResolve from 'rollup-plugin-node-resolve';
import json from 'rollup-plugin-json';
import { uglify } from 'rollup-plugin-uglify';

const banner = [
  '/*!',
  ` * then-ajax v${pkg.version}`,
  ` * (c) 2019 - ${new Date().getFullYear()} jackness`,
  ' * Released under the MIT License.',
  ' */'
].join('\n');

export default {
  input: './src/then-ajax.ts',
  output: [{
    file: './output/then-ajax.js',
    format: 'cjs',
    banner: banner,
    exports: 'named'
  }],
  plugins: [
    nodeResolve({jsnext: true}),
    commonjs(),
    json(),
    typescript({
      typescript: require('typescript')
    }),
    (process.env.NODE_ENV === 'production' && uglify())
  ]
};


src/then-ajax.ts
import { AjaxOptions as _AjaxOptions, ajax as _ajax } from './ajax';

export const ajax = _ajax
export interface AjaxOptions extends _AjaxOptions {

}

export function thenAjax(op: AjaxOptions): Promise<any> {
  return new Promise((resolve, reject) => {
    let isTimeout = false;
    let timeoutKey: any;
    if (op.timeout) {
      timeoutKey = setTimeout(() => {
        isTimeout = true;
        reject(new Error('timeout'));
      }, op.timeout);
    }
    const iParam = Object.assign(op, {
      success(data: any) {
        if (isTimeout) {
          return;
        }
        clearTimeout(timeoutKey);
        resolve(data);
      },
      error(er: Error) {
        reject(er);
      }
    });
    ajax(iParam);
  });
}

export function thenGet(url: string, data?: any, options?: AjaxOptions) {
  return thenAjax(Object.assign({
    data,
    dataType: 'json',
    methods: 'get',
    url
  }, options));
}

export function thenPost(url: string, data?: any, options?: AjaxOptions) {
  return thenAjax(Object.assign({
    data,
    dataType: 'json',
    methods: 'post',
    url
  }, options));
}

export function thenJsonp(url: string, data?: any, options?: AjaxOptions) {
  return thenAjax(Object.assign({
    data,
    dataType: 'jsonp',
    methods: 'get',
    url
  }, options));
}


output/then-ajax.js
"use strict";function encodeData(t,e,n){var o,a=[],r=e;return o=n?n+"["+t+"]":t,"object"==typeof e&&null!==e?a=a.concat(setObjData(e,t)):(o=encodeURIComponent(o),r=encodeURIComponent(r),a.push(o+"="+r)),a}function setObjData(n,o){var a=[];return"[object Array]"===Object.prototype.toString.call(n)?n.forEach(function(t,e){a=a.concat(encodeData("object"==typeof t?""+e:"",t,o))}):"[object Object]"===Object.prototype.toString.call(n)&&Object.keys(n).forEach(function(t){var e=n[t];a=a.concat(encodeData(t,e,o))}),a}function setStrData(t){var e=t.split("&");return e.forEach(function(t){var e=t.split("=");encodeURIComponent(e[0]),encodeURIComponent(e[1])}),e}function ajax(t){var e,n,o,a=t.url||"",r=(t.methods||"get").toLowerCase(),s=t.data||null,c=t.contentType||"",u=t.dataType||"",i=void 0===t.async||t.async,p=t.timeout,d=t.before||function(){return null},f=t.error||function(){return null},j=t.success||function(){return null},l=t.jsonp||"callback",m=t.jsonpcb||"",h=!1,x=null,b=null;function y(){void 0!==p&&(x=setTimeout(function(){"jsonp"===u||(h=!0,b&&b.abort())},p))}s&&("string"==typeof s?s=setStrData(s):"object"==typeof s&&(s=setObjData(s)),s=s.join("&").replace("/%20/g","+"),"get"!==r&&"jsonp"!==u||(a+=-1<a.indexOf("?")?-1<a.indexOf("=")?"&"+s:s:"?"+s)),d(),"jsonp"===u?(n=document.createElement("script"),o="JSONP_"+((new Date).getTime()+Math.round(1e3*Math.random())),m&&(o=m),window[o]=function(t){clearTimeout(x),document.body.removeChild(n),j(t)},n.src=a+(-1<a.indexOf("?")?"&":"?")+l+"="+o,n.type="text/javascript",document.body.appendChild(n)):((b=window.XMLHttpRequest?new XMLHttpRequest:(["Microsoft","msxm3","msxml2","msxml1"].forEach(function(t){try{e=new ActiveXObject(t+".XMLHTTP")}catch(t){}}),e)).open(r,a,i),"post"!==r||c?c&&b.setRequestHeader("Content-Type",c):b.setRequestHeader("Content-Type","application/x-www-form-urlencoded;charset=UTF-8"),b.onreadystatechange=function(){if(4===b.readyState){if(void 0!==p){if(h)return;clearTimeout(x)}200<=b.status&&b.status<300||304===b.status?j("json"===u?JSON.parse(b.responseText):b.responseText):f(b.status,b.statusText)}},b.send("get"===r?null:s)),y()}Object.defineProperty(exports,"__esModule",{value:!0});var ajax$1=ajax;function thenAjax(r){return new Promise(function(e,n){var o,a=!1;r.timeout&&(o=setTimeout(function(){a=!0,n(new Error("timeout"))},r.timeout));var t=Object.assign(r,{success:function(t){a||(clearTimeout(o),e(t))},error:function(t){n(t)}});ajax$1(t)})}function thenGet(t,e,n){return thenAjax(Object.assign({data:e,dataType:"json",methods:"get",url:t},n))}function thenPost(t,e,n){return thenAjax(Object.assign({data:e,dataType:"json",methods:"post",url:t},n))}function thenJsonp(t,e,n){return thenAjax(Object.assign({data:e,dataType:"jsonp",methods:"get",url:t},n))}exports.ajax=ajax$1,exports.thenAjax=thenAjax,exports.thenGet=thenGet,exports.thenJsonp=thenJsonp,exports.thenPost=thenPost;


src/ajax.ts
type callback = (...args: any[]) => any;
export interface AjaxOptions {
  data?: any;
  url?: string;
  methods?: string;
  contentType?: string;
  dataType?: string;
  async?: boolean;
  timeout?: number;
  before?: callback;
  error?: callback;
  success?: callback;
  jsonp?: string;
  jsonpcb?: string;
  withCredentials?: boolean;
}

function encodeData(name: string, value: any, parentName?: string) {
  let items: any[] = [];
  let iName: string;
  let iValue: any = value;
  if (parentName) {
    iName = `${parentName}[${name}]`;
  } else {
    iName = name;
  }

  if (typeof value === 'object' && value !== null) {
    items = items.concat(setObjData(value, name));
  } else {
    iName = encodeURIComponent(iName);
    iValue = encodeURIComponent(iValue);
    items.push(`${iName}=${iValue}`);
  }
  return items;
}

function setObjData(data: any, parentName?: string) {
  let arr: any[] = [];
  if (Object.prototype.toString.call(data) === '[object Array]') {
    data.forEach((value: any, i: number) => {
      arr = arr.concat(encodeData(
        typeof value === 'object' ? `${i}` : '',
        value,
        parentName
      ));
    });
  } else if (Object.prototype.toString.call(data) === '[object Object]') {
    Object.keys(data).forEach((key) => {
      const value = data[key];
      arr = arr.concat(encodeData(key, value, parentName));
    });
  }
  return arr;
}

// 设置字符串的遍码，字符串的格式为：a=1&b=2;
function setStrData(data: string) {
  const arr = data.split('&');
  const r = [];
  arr.forEach((item) => {
    const iArr = item.split('=');
    const name = encodeURIComponent(iArr[0]);
    const value = encodeURIComponent(iArr[1]);
    r.push(`${name}=${value}`);
  });
  return arr;
}

export function ajax(options: AjaxOptions) {
  let url = options.url || '';
  const type = (options.methods || 'get').toLowerCase();
  let data = options.data || null;
  const contentType = options.contentType || '';
  const dataType = options.dataType || '';
  const async = options.async === undefined ? true : options.async;
  const timeOut = options.timeout;
  const before = options.before || (() => null);
  const error = options.error || (() => null);
  const success = options.success || (() => null);
  const jsonp = options.jsonp || 'callback';
  const jsonpcb = options.jsonpcb || '';

  let timeoutBool = false;
  let timeoutFlag: any = null;
  let xhr: any = null;

  function setData() {
    if (data) {
      if (typeof data === 'string') {
        data = setStrData(data);
      } else if (typeof data === 'object') {
        data = setObjData(data);
      }
      data = data.join('&').replace('/%20/g', '+');
      // 若是使用get方法或JSONP，则手动添加到URL中
      if (type === 'get' || dataType === 'jsonp') {
        url += url.indexOf('?') > -1 ? (url.indexOf('=') > -1 ? '&' + data : data) : '?' + data;
      }
    }
  }

  // JSONP
  function createJsonp() {
    const script = document.createElement('script');
    const timeName = new Date().getTime() + Math.round(Math.random() * 1000);

    let callbackName = `JSONP_${timeName}`;
    if (jsonpcb) {
      callbackName = jsonpcb;
    }

    window[callbackName] = (iData: any) => {
      clearTimeout(timeoutFlag);
      document.body.removeChild(script);
      success(iData);
    };
    script.src = `${url}${url.indexOf('?') > -1 ? '&' : '?'}${jsonp}=${callbackName}`;
    script.type = 'text/javascript';
    document.body.appendChild(script);
    setTime(callbackName, script);
  }

  // XHR
  function createXHR() {
    // 由于IE6的XMLHttpRequest对象是通过MSXML库中的一个ActiveX对象实现的。
    // 所以创建XHR对象，需要在这里做兼容处理。
    function getXHR() {
      if (window.XMLHttpRequest) {
        return new XMLHttpRequest();
      } else {
        let r: any;
        // 遍历IE中不同版本的ActiveX对象
        const versions = ['Microsoft', 'msxm3', 'msxml2', 'msxml1'];
        versions.forEach((ver) => {
          try {
            const version = `${ver}.XMLHTTP`;
            r = new ActiveXObject(version);
          } catch (er) {
            // empty
          }
        });
        return r;
      }
    }
    // 创建对象。
    xhr = getXHR();
    xhr.open(type, url, async);
    // 设置请求头
    if (type === 'post' && !contentType) {
      // 若是post提交，则设置content-Type 为application/x-www-four-urlencoded
      xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded;charset=UTF-8');
    } else if (contentType) {
      xhr.setRequestHeader('Content-Type', contentType);
    }
    // 添加监听
    xhr.onreadystatechange = () => {
      if (xhr.readyState === 4) {
        if (timeOut !== undefined) {
          // 由于执行abort()方法后，有可能触发onreadystatechange事件，
          // 所以设置一个timeout_bool标识，来忽略中止触发的事件。
          if (timeoutBool) {
            return;
          }
          clearTimeout(timeoutFlag);
        }
        if ((xhr.status >= 200 && xhr.status < 300) || xhr.status === 304) {
          if (dataType === 'json') {
            success(JSON.parse(xhr.responseText));
          } else {
            success(xhr.responseText);
          }
        } else {
          error(xhr.status, xhr.statusText);
        }
      }
    };
    // 发送请求
    xhr.send(type === 'get' ? null : data);
    setTime(); // 请求超时
  }

  // 设置请求超时
  function setTime(callbackName?: string, script?: Element) {
    if (timeOut !== undefined) {
      timeoutFlag = setTimeout(() => {
        if (dataType === 'jsonp') {
          if (callbackName && script) {
            // delete window[callbackName];
            // document.body.removeChild(script);
          }
        } else {
          timeoutBool = true;
          if (xhr) {
            xhr.abort();
          }
        }
      }, timeOut);
    }
  }

  setData();
  before();
  if (dataType === 'jsonp') {
    createJsonp();
  } else {
    createXHR();
  }
}

